### PR TITLE
blog post: build notes ai summarization 

### PR DESCRIPTION
**Pull request by KyleAMathews**

### CODE REVIEW

File: src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+Which is awesome! I love having the full power of SQL to query data + built-in reactivity for real-time updates. It’s really everything I’ve ever wanted for client-side data.
+
+But there’s two issues before your reactive query can go to work. First is ensuring the data you want to query is synced. And the second is pre-running the query so the new route can immediately render.
+
+
+### Route Syncing
+
+The simplest way to build a local-first app is to just sync upfront all data but this gets slow as the data size grows. So just like with code splitting, you can split data syncing along route boundaries so the user waits for only the minimal amount of data to be synced.
+
+It’d be really rare to want an entire database synced to each client. So instead you can specify which tables and even which subsets of tables e.g. only load the latest 10 notifications.
```
Comment (lorensr) [inline src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md:39]:
```suggestion
It’d be really rare to want an entire database synced to each client. So instead you can specify which tables and even which subsets of tables — e.g. only load the latest 10 notifications.
```

File: src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+          shapeMetadata.state === `SYNCED`,
+      },
+    ])
+    return null
+  },
+}
+```
+
+
+As these are all functions, you could reuse common shapes across routes e.g. user and org data. You can also flexibly say what “done” is e.g. for some routes, somewhat stale data is fine so don’t wait on loading for the freshest data to be loaded from the server.
```
Comment (lorensr) [inline src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md:83]:
```suggestion
As these are all functions, you could reuse common shapes across routes e.g. user and org data. You can also flexibly say what “done” is—e.g. for some routes, somewhat stale data is fine, so don’t wait on loading for the freshest data to be loaded from the server.
```

File: src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+
+### Pre-running queries
+
+The other problem I ran into is that while local SQLite queries run very fast, they are async and take a bit of time so on transitioning to a route, there’s some number of frames where nothing renders. So while fast, it causes a very noticeable blink in the UI.
```
Comment (lorensr) [inline src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md:88]:
```suggestion
The other problem I ran into is that while local SQLite queries run very fast, they are async and take a bit of time, so on transitioning to a route, there’s some number of frames where nothing renders. So while fast, it causes a very noticeable blink in the UI.
```

File: src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+## Real-timey bits
+
+This was a pretty stock standard app. The only thing I did which really took advantage of ElectricSQL superpowers was the progress bar for indicating progress of the summarization.
+
+<video controls width="500">
+  <source src="./progress-bar.mp4" type="video/mp4" />
+</video>
+
+As ElectricSQL gives you a full-stack reactive data layer for free — it was quite easy. A progress bar is just a value between 0 & 1 i.e. what percentage of the work has been done. The question for this sort of long-running job is how to get that value from the backend doing the work to the frontend for the user to see. Normally it’d be some sort of string and glue hacky setup where the backend pushes events over websockets which the client code then has custom code to intercept and direct to the component through a global data store of some sort.
```
Comment (lorensr) [inline src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md:132]:
```suggestion
As ElectricSQL gives you a full-stack reactive data layer for free — it was quite easy. A progress bar is just a value between 0 & 1, i.e., what percentage of the work has been done. The question for this sort of long-running job is how to get that value from the backend doing the work to the frontend for the user to see. Normally it’d be some sort of string and glue hacky setup where the backend pushes events over websockets which the client code then has custom code to intercept and direct to the component through a global data store of some sort.
```

File: src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+But none of that here. I have a `youtube_video` table for each video which has a `progress` column. The backend just keeps writing new values to that as summarization calls finish and the new value is synced directly into the Video component.
```
Comment (lorensr) [inline src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md:134]:
```suggestion
But none of that here. I have a `youtube_video` table for each video which has a `progress` column. The backend just keeps writing new values to that as summarization calls finish, and the new value is synced directly into the Video component.
```

File: src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+An ElectricSQL app has three components
+
+* The client — which works the same as any other web/mobile app — I used React/react-router and am hosting on Vercel.
+* Postgres — any of the 1000s of Postgres hosting options (as long as it supports Logical Replication)
+* The Electric sync service — it runs anywhere Docker runs and the Samurize instance is currently using ~325 MB of ram.
+
+Postgres hosting gives you a lot of nice things like backups and a dashboard. But for a side project like this, I don’t care about losing data really and psql works for a dashboard. So I got a DigitalOcean VM and created a Docker Compose file with containers for Postgres and Electric. I needed a way to terminate SSL for an encrypted websocket connection and after looking at Nginx and Caddy, decided on Caddy as it’s extremely easy to use — you just tell it what domain it’s handling and it automatically gets an SSL cert from Let’s Encrypt. Ngnix works well too but it’s much more manual to get certs. I was pleasantly surprised to see Caddy is only using ~13 MB of ram w/ my extremely low traffic side project.
```
Comment (lorensr) [inline src/pages/2023-12-13--build-notes-ai-powered-youtube-summarizer/index.md:150]:
```suggestion
Postgres hosting gives you a lot of nice things like backups and a dashboard. But for a side project like this, I don’t care about losing data really, and psql works for a dashboard. So I got a DigitalOcean VM and created a Docker Compose file with containers for Postgres and Electric. I needed a way to terminate SSL for an encrypted websocket connection and after looking at Nginx and Caddy, decided on Caddy as it’s extremely easy to use — you just tell it what domain it’s handling and it automatically gets an SSL cert from Let’s Encrypt. Ngnix works well too but it’s much more manual to get certs. I was pleasantly surprised to see Caddy is only using ~13 MB of ram w/ my extremely low traffic side project.
```

**Review by lorensr — COMMENTED** [review]
sounds like Meteor's data system! happy it's spreading ☺️

**Comment by KyleAMathews** [comment]
It is!

**Pull request merged**